### DISCUSSION TITLE
Backlog inter-command delay (SetOption34) is not forced

### DISCUSSION THREAD

**Discussion by jozseflkiss** [arendst/Tasmota#15393] [Q&A] [answered]
Hi,

As far as I see the backlog inter-command delay is not forced when receiving subsequent backlog commands (e.g. via MQTT). 
From the source code, I think the TasmotaGlobal.backlog_timer = millis() at the end of CmndBacklog function possibly resets the already defined backlog_timer (which has been postponed by SetOption34 value, Settings->param[P_BACKLOG_DELAY]). Do I miss something?
![SetOption34](https://user-images.githubusercontent.com/50739844/163770488-14cfdf24-f75b-4f36-a13f-765bbcc608b6.jpg)
?

**Comment by sfromis** [comment, marked as answer]
Tasmota does not have reentrant multithreading. Do not expect that multiple commands at the same time (including the backlog command as such) will proceed undisturbed by each other. If you need more reliable timing, use `RuleTimer`

**Reply by jozseflkiss** [reply]
Thanks for the input! I have a Sonoff RF Bridge which controls shutters. I want to open all the shutters in the house, so multiple MQTT commands are sent to the Sonoff at the same time (cover group in HA), however based on my experience I need to delay the RfRaw commands to have all the shutter work as desired. Is this achievable with RuleTimer? How?

**Reply by sfromis** [reply]
Here's a simple example of how RuleTimer can be used. https://tasmota.github.io/docs/Rules/#auto-off-motion-sense-switch
If you feel a need to stay with the resolution of SetOption34, you could have it all in one long backlog sequence, or via one `Event` command activate a rule to use a RuleTimer to control the rest before firing off the time-sensitive commands. Just make sure to execute no other command while a Backlog sequence is running.

**Reply by jozseflkiss** [reply]
Thanks! I checked the documentation and I think the RuleTimer cannot be used in my case because I cannot "merge" commands from different MQTT messages with Rules. I have multiple, simultaneously fired MQTT messages with RfRaw commands in it, which need to be composed as a common command queue, and commands in it should be executed one after another with a small delay between them. Am I right this cannot be done with RuleTimer?

**Reply by sfromis** [reply]
Using rules, it should still be doable to "merge" commands into one dynamic queue, but it gets more involved, like putting commands into vars to manage a dynamic queue, picking one at a time at intervals.

Still, I'd suggest doing something about the bad scheme of wanting to fire off multiple simultaneous MQTT messages, things gets simpler if you can construct one command sequence for one shot.

**Reply by jozseflkiss** [reply]
Finally I created a custom build: in `void CmndBacklog(void)` function I wait for the current timer if exists, so I changed the `TasmotaGlobal.backlog_timer = millis();` line to `TasmotaGlobal.backlog_timer = MAX (TasmotaGlobal.backlog_timer, millis());`. It seems to have solved the `SetOption34` and the Delay problems. Maybe it's worth rethinking to port it to main. Anyway, thanks for everything! Fantastic firmware!!!

**Reply by dhoffend** [reply]
I run into the same problem. I see two problems

1) the rfraw commands is a non-blocking command. So the main backlog/command loop will not know when the command has been finished. Tasmota tries to execute the next command (in our case rfraw) while the radio firmware is likely ignoring the command when the radio device is currently in use.

2) I've seen similar issues with the mqtt + backlog delay problem. When I create a backlog command with 3 rfraw commands (in one backlog job) I see that `SetOption34` is correctly honored, but Still only the first rfraw command will gets executed. due to 255 seems to be not enough pause between commands sometimes, The real problems comes with backlog command inserts from mqtt. the first job gets added and gets directly processed (fire and forget) and at the end the backendloop detects that no other jobs are in the queue and resets the backlog_timer to `millies()` ... Then the next mqtt job gets added and directly executed without taking care of backlog delay of the last run job.

On the other hand ... when 255ms (the max value for `SetOption34`) is not enough between rf commands I see only one option.

Not sure if this works better: Set a lower backlog delay (example `SetOption34 1` to 10ms or so ...) and then always add a `; delay xx` after each rfraw command using Backlog. Maybe this could reduce ignored concurrent sends actions. but it will reduce the number of backlog jobs and I still had issues with (nearly) simultaneous commands from time to time. That's a quest to find the the best job delay + command delay combination

**Reply by sfromis** [reply]
Using `Delay` is not reliable, it can be cut short by things like other commands.
If SetOption34 max of 255 is not enough, you can have multiple intervals by inserting "dummy" commands in the Backlog sequence.
